Allow provider-specific data types (#529)
Materials-Consortia/OPTIMADE: optimade.rst

context-independent types that are assumed to have some form of
representation in all contexts. They are as follows:

> - Basic types: **string**, **integer**, **float**, **boolean**, **timestamp**

Reviewer: ```suggestion
- Basic types: **string**, **integer**, **float**, **boolean**, **timestamp**.
```

context-independent types that are assumed to have some form of
representation in all contexts. They are as follows:

> - Basic types: **string**, **integer**, **float**, **boolean**, **timestamp**, database-provider-specific or definition-provider-specific data type.

Reviewer: The same `database-provider-specific or definition-provider-specific data type` phrase is repeated multiple times throughout the specification. I like that we are being really specific, but maybe we could defined a term somewhere (e.g. `custom data type`) and then refer to it troughout the text?

Author: Fair point, I will check the term definition section to see how well it would fit there.

Reviewer 2: On another point: I think it seems odd to list provider-specific data types under "Basic types". To the extent that we do the separation into Basic vs. list/dictionary on the basis of "contains one thing" vs "many things", it also isn't clear to me why the defined types would have to be seen as the former. I suggest we move the segment about database-specific datatypes below the definitions of the datatypes that are explicitly defined by the standard.

Author: My idea is to allow only those provider-specific data types that can be expressed as strings (e.g., symmetry operators, complex numbers etc). Timestamp type which OPTIMADE already has could fall under the same category of strings with internal semantics. That is why I lumped provider-specific data types together with them. But I agree that they could be split off to a separate segment.

Author: > The same `database-provider-specific or definition-provider-specific data type` phrase is repeated multiple times throughout the specification. I like that we are being really specific, but maybe we could defined a term somewhere (e.g. `custom data type`) and then refer to it troughout the text?

What about calling them "namespace-specific data types"? The specification already has a section "Namespace Prefixes" which is later split into database-specific and definition provider-specific ones, thus I think "namespace" is a right term to use.

Author: I have replaced "database-provider-specific or definition-provider-specific" with "namespace-specific" in 3dd56bb2e500d933df6b2857078cb34ae17636ad.

  A list can be empty, i.e., contain no items.
- **dictionary**: an associative array of **keys** and **values**, where **keys** are pre-determined strings, i.e., for the same entry property, the **keys** remain the same among different entries whereas the **values** change.
  The **values** of a dictionary can be any basic type, list, dictionary, or unknown.

> - Namespace-specific data type.

Reviewer: Maybe it would make sense to add a link here to the part of the specification where "Namespace-specific" data types are defined, e.g. "Type handling and conversions in comparisons"?

Author: Link added, marking as resolved.

- **timestamp** uses a string representation of date and time as defined in `RFC 3339 Internet Date/Time Format <https://tools.ietf.org/html/rfc3339#section-5.6>`__.
- **dictionary** is represented by the JSON object type.
- **unknown** properties are represented by either omitting the property or by a JSON :field-val:`null` value.

> - namespace-specific data types use string representations.

Reviewer: ```suggestion
- Namespace-specific data types use string representations.
```

  (Note that at the end of the string value above the four final backslashes represent the two terminal backslashes in the value, and the final double quote is a terminator, it is not escaped.)

  String value tokens are also used to represent **timestamps** in form of the `RFC 3339 Internet Date/Time Format <https://tools.ietf.org/html/rfc3339#section-5.6>`__.

>   String value tokens as well are used to represent namespace-specific data types.

Reviewer: Maybe this could be made into a small list, i.e.:

String value tokens are also used to represent:
- **timestamps** in form of the `RFC 3339 Internet Date/Time Format <https://tools.ietf.org/html/rfc3339#section-5.6>`__.
- Namespace-specific data types.

In a comparison with a timestamp property, a string token represents a timestamp value that would result from parsing the string according to RFC 3339 Internet Date/Time Format.
Interpretation failures MUST be reported with error :http-error:`400 Bad Request`.

Database and definition providers MAY introduce custom types, representing them with string lexical tokens both in filters and responses.
It is up to the providers to decide which comparison operators to support and how the comparisons should be performed.

> For example, if a provider intoduces a set-valued property :property:`_exmpl_set`, it may decide to define operator :val:`CONTAINS` so that :filter:`identifier CONTAINS set` is true if :val:`set` is a subset of a property.

Reviewer: ```suggestion
For example, a provider that introduces a set-valued property :property:`_exmpl_set`, may decide to override the :val:`CONTAINS` operator so that :filter:`identifier CONTAINS set` is true when :val:`set` is a subset of a property.
```

In a comparison with a timestamp property, a string token represents a timestamp value that would result from parsing the string according to RFC 3339 Internet Date/Time Format.
Interpretation failures MUST be reported with error :http-error:`400 Bad Request`.

> Namespace providers MAY introduce custom types, representing them with string lexical tokens both in filters and responses.

Reviewer: ```suggestion
Namespace providers MAY introduce namespace-specific data types, representing them with string lexical tokens both in filters and responses.
```

- :field:`x-optimade-type`: String.
  Specifies the OPTIMADE data type for this level of the defined property.

>   MUST be one of :val:`"string"`, :val:`"integer"`, :val:`"float"`, :val:`"boolean"`, :val:`"timestamp"`, :val:`"list"`, :val:`"dictionary"` or the name of a namespace-specific data type that starts with an underscore followed by the provider-specific prefix.

Reviewer 3: Should we add here `blob_raw_bytes` and `pointer_reference` 🤔

Author: We may, but maybe as a separate PR? This one is made specifically for namespace provider-specific data types like SMILES in cheminformatics and the like.

- :field:`x-optimade-type`: String.
  Specifies the OPTIMADE data type for this level of the defined property.

>   MUST be one of :val:`"string"`, :val:`"integer"`, :val:`"float"`, :val:`"boolean"`, :val:`"timestamp"`, :val:`"list"`, :val:`"dictionary"` or a string naming namespace-specific data type starting with provider-specific prefix.

Reviewer: ```suggestion
  MUST be one of :val:`"string"`, :val:`"integer"`, :val:`"float"`, :val:`"boolean"`, :val:`"timestamp"`, :val:`"list"`, :val:`"dictionary"` or the name of a namespace-specific data type that starts with an underscore followed by the provider-specific prefix.
```

Reviewer: Feel free to reword this in a different way than suggested,

>   MUST be one of :val:`"string"`, :val:`"integer"`, :val:`"float"`, :val:`"boolean"`, :val:`"timestamp"`, :val:`"list"`, :val:`"dictionary"` or the name of a namespace-specific data type that starts with an underscore followed by the provider-specific prefix.

Reviewer 2: I guess in the end we will want this info to go into our "standards" definitions (see https://github.com/Materials-Consortia/OPTIMADE/tree/develop/schemas#standards-definitions ). For the OPTIMADE one see here: https://schemas.optimade.org/defs/v1.2/standards/optimade.html 

If forced to handle this today I would put that info into the description field there. Something along the lines of: 

> Description: The Magnetic materials standard node types and properties.
>
> Apart from the supplied property definitions, the Magnetic materials standard defines a namespace-specific data type for the magnetic susceptibility tensor "_mag_sustensor" that allows comparison using < and > to express filters of the magnetic susceptibility magnitude. In more detail, the following operators are valid: 
> ...

BUT, writing this up; even though the magnetic susceptibility tensor example is quite bogus, I realize we may be overly limiting deciding all custom datatypes must be represented as strings? Why not allow it to be expressed as any basic or container datatype? 

But, back to @ml-evs comment: to allow not repeating the definition a lot, maybe this is better?:

> When a namespace-specific data type is used, the human-readable property description should be used to describe its usage and format or to give a reference to where this information is available.

(That reference can then be to the "standards definition" if it is included there).

>   MUST be one of :val:`"string"`, :val:`"integer"`, :val:`"float"`, :val:`"boolean"`, :val:`"timestamp"`, :val:`"list"`, :val:`"dictionary"` or the name of a namespace-specific data type that starts with an underscore followed by the provider-specific prefix.

Author: I think this comment better fits to "Property Definition keys from JSON Schema" section, next to definition of `type`.

>   MUST be one of :val:`"string"`, :val:`"integer"`, :val:`"float"`, :val:`"boolean"`, :val:`"timestamp"`, :val:`"list"`, :val:`"dictionary"` or the name of a namespace-specific data type that starts with an underscore followed by the provider-specific prefix.

Reviewer 2: @ml-evs This is no longer relevant with the latest changes; I 'resolve' this now. If you still see issues, make a new comment.
